### PR TITLE
Fixes #30190 - adds verify checksum

### PR DESCRIPTION
**Pull request by jjeffers**

### CODE REVIEW

File: app/controllers/katello/api/v2/products_bulk_actions_controller.rb [hunk 1]
```diff
@@ -50,6 +50,20 @@ def sync_products
       respond_for_async :resource => task
     end
 
+    api :PUT, "/products/bulk/verify_checksum", N_("Verify checksum for one or more products")
+    param :ids, Array, :desc => N_("List of product ids"), :required => true
+    def verify_checksum_products
+      repairable_products = @products.syncable
+      repairable_roots = RootRepository.where(:product_id => repairable_products).has_url
+
+      repairable_repositories = Katello::Repository.where(:root_id => repairable_roots).in_default_view
```
Comment (jlsherrill) [inline app/controllers/katello/api/v2/products_bulk_actions_controller.rb:59]: you probably wanna filter to only yum repos

Comment (jjeffers) [reply app/controllers/katello/api/v2/products_bulk_actions_controller.rb:59]: So at this point, is pulp_rpm the only content type that will offer the repair api? I wasn't sure if additional content types would implement this?

Comment (jlsherrill) [reply app/controllers/katello/api/v2/products_bulk_actions_controller.rb:59]: thats my understanding, and pulp2 only supports yum as well

**Comment by jlsherrill** [comment]
Overall this looks great!   I can start testing and do a full review if its ready

**Comment by jjeffers** [comment]
> Overall this looks great! I can start testing and do a full review if its ready

Me: squints at "looks great". 

I was pretty sure this looked not-ready-for-primetime!

File: app/controllers/katello/api/v2/products_bulk_actions_controller.rb [hunk 2]
```diff
@@ -50,6 +50,21 @@ def sync_products
...
     end
 
+    api :PUT, "/products/bulk/verify_checksum", N_("Verify checksum for one or more products")
+    param :ids, Array, :desc => N_("List of product ids"), :required => true
+    def verify_checksum_products
+      repairable_products = @products.syncable
+      repairable_roots = RootRepository.where(:product_id => repairable_products).
+        where(:content_type => ::Katello::Repository::YUM_TYPE).has_url
+
+      repairable_repositories = Katello::Repository.where(:root_id => repairable_roots)
```
Comment (jlsherrill) [inline app/controllers/katello/api/v2/products_bulk_actions_controller.rb:60]: we may want to only repair the Library instance repos of a product.  Repairing all repos across all content views could take a very long time, and likely contains the same content thats in the Library repo

File: app/controllers/katello/api/v2/repositories_controller.rb [hunk 3]
```diff
@@ -15,7 +15,7 @@ class Api::V2::RepositoriesController < Api::V2::ApiController # rubocop:disable
     before_action :find_organization_from_product, :only => [:create]
     before_action :find_repository, :only => [:show, :update, :destroy, :sync, :export,
                                               :remove_content, :upload_content, :republish,
-                                              :import_uploads, :gpg_key_content]
+                                              :import_uploads, :gpg_key_conten, :verify_checksum]
```
Comment (jlsherrill) [inline app/controllers/katello/api/v2/repositories_controller.rb:18]: typo here on 'content'

File: app/services/katello/pulp3/repository/yum.rb [hunk 4]
```diff
@@ -112,6 +112,11 @@ def copy_units(source_repository, content_unit_hrefs, dependency_solving, dest_b
           tasks
         end
 
+        def repair(repository_version_href)
+          data = PulpRpmClient::RepositoryVersion.new
+          api.repository_versions_api.repair(repository_version_href, data)
```
Comment (jlsherrill) [inline app/services/katello/pulp3/repository/yum.rb:117]: out of curiosity, why do you have to pass a new/blank instance of  PulpRpmClient::RepositoryVersion  ?

File: app/controllers/katello/api/v2/repositories_controller.rb (hunk 3 above)
Comment (jjeffers) [reply app/controllers/katello/api/v2/repositories_controller.rb:18]: yikes!

File: app/services/katello/pulp3/repository/yum.rb (hunk 4 above)
Comment (jjeffers) [reply app/services/katello/pulp3/repository/yum.rb:117]: The RepositoryVersion is required, but curiously the only data element is the `base_version` value, which is optional. Without the RepositoryVersion the pulp_rpm_client throws an exception, for example `wrong number of arguments (given 1, expected 2..3)`

**Comment by jlsherrill** [comment]
The task status looks like this:

image

Would it be possible to add a presenter to show this a bit better?

**Comment by jlsherrill** [comment]
Trying to do the workflow with pulp3 from the 'advanced sync' page (aka which uses the sync api), the task errors, and it looks like it still is calling some pulp2 actions

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/views/products.html [hunk 5]
```diff
@@ -36,6 +36,12 @@ <h2 translate>Products</h2>
           </a>
         </li>
 
+        <li role="menuitem" ng-show="permitted('edit_products')" ng-class="{disabled: table.numSelected === 0}">
+          <a ng-click="verifyChecksumProducts()" disable-link="table.numSelected === 0" translate>
+            Verify Checksum
```
Comment (jlsherrill) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/views/products.html:41]: "Verify Content Checksums"  Might be more clear what its doing

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/details/repositories/details/views/repository-details.html [hunk 6]
```diff
@@ -32,6 +32,12 @@ <h2 translate>{{ repository.name }}</h2>
           </a>
         </li>
 
+        <li role="menuitem" ng-hide="hideSyncButton(repository, true)" ng-class="{disabled: disableSyncLink()}">
+          <a ng-click="verifyChecksum(repository)" disable-link="disableSyncLink()" translate>
+            Verify Checksum
```
Comment (jlsherrill) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/details/repositories/details/views/repository-details.html:37]: same comment here, i think i like 'Verify Content Checksums'

**Comment by jlsherrill** [comment]
overall looks good and testing with pulp3 and pulp2 shows correct results.  I think you'll need to add this new controller action to permission_creator.rb  though, in order for it to be protected properly

**Comment by jlsherrill** [comment]
looks like rubocop, angular tests, and ruby test failures

**Comment by jlsherrill** [comment]
[test katello]

**Comment by jjeffers** [comment]
[test katello]

**Pull request merged**